### PR TITLE
Implement clear scores on beatmap

### LINKED ISSUE #3941: "Clear local scores" option doesn't work
**Describe your problem:** The problem is with the "Clear local scores" button: it clears nothing.

**Screenshots or videos showing encountered issue:** 
https://streamable.com/r2m68

**osu!lazer version:** 2018.1227.1.0

**Logs:** Unnecessary

### PR DESCRIPTION
**Pull request by m2307**
Implementation for "Clear local scores" on beatmaps

- Closes #3941

### CODE REVIEW

File: osu.Game/Scoring/ScoreManager.cs [hunk 1]
```diff
@@ -55,9 +55,7 @@ protected override ScoreInfo CreateModel(ArchiveReader archive)
 
         public Score GetScore(ScoreInfo score) => new LegacyDatabasedScore(score, rulesets, beatmaps, Files.Store);
 
-        public List<ScoreInfo> GetAllUsableScores() => ModelStore.ConsumableItems.Where(s => !s.DeletePending).ToList();
-
-        public IEnumerable<ScoreInfo> QueryScores(Expression<Func<ScoreInfo, bool>> query) => ModelStore.ConsumableItems.AsNoTracking().Where(query);
+        public IEnumerable<ScoreInfo> GetAllUsableScores() => ModelStore.ConsumableItems.Where(s => !s.DeletePending);
```
Comment (peppy) [inline osu.Game/Scoring/ScoreManager.cs]: why this change?

File: osu.Game/Screens/Select/BeatmapClearScoresDialog.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using System.Linq;
+
+namespace osu.Game.Screens.Select
+{
+    public class BeatmapClearScoresDialog : PopupDialog
+    {
+        private ScoreManager manager;
+
+        [BackgroundDependencyLoader]
+        private void load(ScoreManager beatmapManager)
```
Comment (peppy) [inline osu.Game/Screens/Select/BeatmapClearScoresDialog.cs]: beatmapManager?

File: osu.Game/Screens/Select/BeatmapClearScoresDialog.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            manager = beatmapManager;
+        }
+
+        public BeatmapClearScoresDialog(BeatmapSetInfo beatmap, IEnumerable<ScoreInfo> scores, Action refresh)
```
Comment (peppy) [inline osu.Game/Screens/Select/BeatmapClearScoresDialog.cs]: why do you pass in the score rather than just getting them here?

File: osu.Game/Screens/Select/SongSelect.cs [hunk 4]
```diff
@@ -625,6 +625,15 @@ private void delete(BeatmapSetInfo beatmap)
             dialogOverlay?.Push(new BeatmapDeleteDialog(beatmap));
         }
 
+        private void clearScores(BeatmapSetInfo beatmap)
+        {
+            if (beatmap == null || beatmap.ID <= 0) return;
```
Comment (peppy) [inline osu.Game/Screens/Select/SongSelect.cs]: you aren't checking if the user is looking at local scores here. there's a high chance you're incorrectly passing online scores from the leaderboard to be deleted.

**Review by peppy — DISMISSED** [review]

File: osu.Game/Screens/Select/BeatmapClearScoresDialog.cs (hunk 3 above)
Comment (m2307) [reply osu.Game/Screens/Select/BeatmapClearScoresDialog.cs]: Isn't passing them through rather than getting them here again more performant?

File: osu.Game/Screens/Select/SongSelect.cs (hunk 4 above)
Comment (m2307) [reply osu.Game/Screens/Select/SongSelect.cs]: Added this in https://github.com/ppy/osu/pull/4000/commits/472325b885b4ba54e2b9c4a7383c0d2997298d51

File: osu.Game/Screens/Select/BeatmapClearScoresDialog.cs (hunk 2 above)
Comment (m2307) [reply osu.Game/Screens/Select/BeatmapClearScoresDialog.cs]: Fixed in https://github.com/ppy/osu/pull/4000/commits/562c9e56fbda6c85dc2ab98402296ec461232eeb

File: osu.Game/Screens/Select/BeatmapClearScoresDialog.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapClearScoresDialog.cs]: in a scenario like this performance is not even a consideration

File: osu.Game/Scoring/ScoreManager.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Scoring/ScoreManager.cs]: Couldyou not use `QueryScores` below? These methods generally should match on all manager classes (we probably want an interface for this going forward).

File: osu.Game/Screens/Select/BeatmapClearScoresDialog.cs (hunk 3 above)
Comment (m2307) [reply osu.Game/Screens/Select/BeatmapClearScoresDialog.cs]: Resolved in https://github.com/ppy/osu/pull/4000/commits/097062caf78bd7985fa4d1af9b675743191a6671

File: osu.Game/Scoring/ScoreManager.cs (hunk 1 above)
Comment (m2307) [reply osu.Game/Scoring/ScoreManager.cs]: Resolved in https://github.com/ppy/osu/pull/4000/commits/097062caf78bd7985fa4d1af9b675743191a6671

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Select/SongSelect.cs [hunk 5]
```diff
@@ -625,6 +626,17 @@ private void delete(BeatmapSetInfo beatmap)
             dialogOverlay?.Push(new BeatmapDeleteDialog(beatmap));
         }
 
+        private void clearScores(BeatmapInfo beatmap)
+        {
+            if (BeatmapDetails.Leaderboard.Scope != BeatmapLeaderboardScope.Local) return;
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Select/SongSelect.cs]: agree with this one, but leaving the second check below.

**Comment by peppy** [comment]
@smoogipoo requesting re-review.

- Fixed database relying on cascade deletion to delete files.
- Centralised this logic so we don't have to manually fiddle with includes.

File: osu.Game/Screens/Select/SongSelect.cs [hunk 6]
```diff
@@ -625,6 +626,17 @@ private void delete(BeatmapSetInfo beatmap)
...
+
+            if (beatmap == null || beatmap.ID <= 0) return;
+
+            if (BeatmapDetails.Leaderboard.Scores == null || !BeatmapDetails.Leaderboard.Scores.Any()) return;
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Screens/Select/SongSelect.cs]: This was marked as resolved, why @peppy ? This is going to break if the leaderboard isn't showing local scores?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — DISMISSED** [review]
Leaderboard should be reloaded after deletion.

**Review by smoogipoo — APPROVED** [review]
